### PR TITLE
Feat/datumaro annotations

### PR DESCRIPTION
**Pull request by Mark-Niemeyer**
`labels_bb` and `labels_general` are merged to `labels` + datumaro-json. Only the label and the instance uuid are still explicit. All annotations types (bounding boxes, semantic segmantation, key points etc.) must be stored in the datumaro-json.

@J-Bracke: This is only the server code. There is no importer/converter from the raw datumaro-json yet.

still open:

- [x] python gRPC examples aren't updated yet. Wait for the merge of  feat/integration-tests-pyscripts and update afterwards

### CODE REVIEW

File: seerep_msgs/protos/label.proto [hunk 1]
```diff
@@ -5,5 +5,7 @@ package seerep.pb;
 message Label
 {
   string label = 1;
-  float confidence = 2;
+  int32 labelIdDatumaro = 2;
+  string instanceUuid = 3;
+  int32 instanceIdDatumaro = 4;
```
Comment (jarkenau) [inline seerep_msgs/protos/label.proto:10]: It's generally not recommended to use `int32` or `int64` as for IDs https://protobuf.dev/programming-guides/api/#integer-field-for-id, but instead use a `string`. However, I can see that it's very unlikely that we will get IDs larger than $2^{32}-1$ in the datumaro json.

File: seerep_com/fbs/point_cloud_service.fbs [hunk 2]
```diff
@@ -1,7 +1,8 @@
 
 include "point_cloud_2.fbs";
 include "query.fbs";
-include "boundingboxes_labeled_stamped.fbs";
+include "label_category.fbs";
```
Comment (jarkenau) [inline seerep_com/fbs/point_cloud_service.fbs:4]: Unnecessary import?

File: seerep_srv/seerep_server/src/fb_image_service.cpp [hunk 3]
```diff
@@ -147,36 +147,36 @@ grpc::Status FbImageService::TransferImage(grpc::ServerContext* context,
...
+  while (reader->Read(&labelMsg))
   {
-    BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::info) << "received BoundingBoxes2DLabeledStamped... ";
-    auto bbslabeled = bbsMsg.GetRoot();
+    BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::info) << "received label...";
+    auto label = labelMsg.GetRoot();
 
-    std::string uuidProject = bbslabeled->header()->uuid_project()->str();
+    std::string uuidProject = label->projectUuid()->str();
     if (uuidProject.empty())
```
Comment (jarkenau) [inline seerep_srv/seerep_server/src/fb_image_service.cpp:165]:
```suggestion
    const std::string uuidProject = label->projectUuid()->str();
```

File: seerep_srv/seerep_server/src/fb_image_service.cpp [hunk 4]
```diff
@@ -147,36 +147,36 @@ grpc::Status FbImageService::TransferImage(grpc::ServerContext* context,
...
+  flatbuffers::grpc::Message<seerep::fb::DatasetUuidLabel> labelMsg;
+  while (reader->Read(&labelMsg))
   {
-    BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::info) << "received BoundingBoxes2DLabeledStamped... ";
-    auto bbslabeled = bbsMsg.GetRoot();
+    BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::info) << "received label...";
+    auto label = labelMsg.GetRoot();
 
-    std::string uuidProject = bbslabeled->header()->uuid_project()->str();
+    std::string uuidProject = label->projectUuid()->str();
```
Comment (jarkenau) [inline seerep_srv/seerep_server/src/fb_image_service.cpp:164]:
```suggestion
    const std::string uuidProject = label->projectUuid()->str();
```

Comment (jarkenau) [inline seerep_srv/seerep_server/src/fb_image_service.cpp:164]:
```suggestion
    const std::string uuidProject = label->projectUuid()->str();
```

File: seerep_srv/seerep_core_fb/src/core_fb_image.cpp [hunk 5]
```diff
@@ -66,47 +66,49 @@ boost::uuids::uuid CoreFbImage::addData(const seerep::fb::Image& img)
...
         catch (std::runtime_error const& e)
         {
-          labelWithInstance.uuidInstance = boost::uuids::nil_uuid();
+          label.uuidInstance = boost::uuids::nil_uuid();
         }
-        labelsWithInstances.push_back(labelWithInstance);
+        labelVector.push_back(label);
       }
-      labelWithInstancePerCategory.emplace(bbWithCategory->category()->c_str(), labelsWithInstances);
+      labelPerCategory.emplace(labelCategory->category()->c_str(), labelVector);
```
Comment (jarkenau) [inline seerep_srv/seerep_core_fb/src/core_fb_image.cpp:103]: Unnecessary implicit conversion.
```suggestion
      labelPerCategory.emplace(labelCategory->category()->str(), labelVector);
```

File: seerep_srv/seerep_core_fb/include/seerep_core_fb/core_fb_conversion.h [hunk 6]
```diff
@@ -46,7 +47,7 @@
 
 namespace seerep_core_fb
 {
-using LabelWithInstanceVec = std::vector<seerep_core_msgs::LabelWithInstance>;
+using LabelVec = std::vector<seerep_core_msgs::Label>;
```
Comment (jarkenau) [inline seerep_srv/seerep_core_fb/include/seerep_core_fb/core_fb_conversion.h:50]: https://github.com/agri-gaia/seerep/pull/354#discussion_r1422669544 ?

File: seerep_srv/seerep_core_fb/src/core_fb_conversion.cpp [hunk 7]
```diff
@@ -765,4 +682,33 @@ std::vector<seerep_core_msgs::Datatype> CoreFbConversion::fromFbDatatypeVector(c
...
+//       labelCategory.labelsIdDatumaro = labelsIdDatumaro;
+//       labelCategory.instances = instances;
+//       labelCategory.instancesIdDatumaro = instancesIdDatumaro;
+//       labelCategory.datumaroJson = labelsCategory->datumaroJson()->c_str();
+//       labelCategoryVector.push_back(labelCategory);
+//     }
+//   }
+//   return labelCategoryVector;
+// }
 }  // namespace seerep_core_fb
```
Comment (jarkenau) [inline seerep_srv/seerep_core_fb/src/core_fb_conversion.cpp:714]: Remove this?

File: seerep_srv/seerep_core_fb/src/core_fb_image.cpp [hunk 8]
```diff
@@ -66,47 +66,49 @@ boost::uuids::uuid CoreFbImage::addData(const seerep::fb::Image& img)
...
-  boost::uuids::uuid uuidMsg = gen(boundingBoxes2dlabeled.header()->uuid_msgs()->str());
-  boost::uuids::uuid uuidProject = gen(boundingBoxes2dlabeled.header()->uuid_project()->str());
+  boost::uuids::uuid uuidMsg = gen(datasetUuidLabel.datasetUuid()->str());
+  boost::uuids::uuid uuidProject = gen(datasetUuidLabel.projectUuid()->str());
 
   auto hdf5io = CoreFbGeneral::getHdf5(uuidProject, m_seerepCore, m_hdf5IoMap);
 
-  hdf5io->writeImageBoundingBox2DLabeled(boost::lexical_cast<std::string>(uuidMsg), boundingBoxes2dlabeled.labels_bb());
+  hdf5io->writeLabelsFb(boost::lexical_cast<std::string>(uuidMsg), datasetUuidLabel.datasetUuid()->str(),
+                        datasetUuidLabel.labels());
```
Comment (jarkenau) [inline seerep_srv/seerep_core_fb/src/core_fb_image.cpp:79]: For one UUID a cast is used, the other is simply extracted from the method parameter.

File: seerep_srv/seerep_core_fb/src/core_fb_pointcloud.cpp [hunk 9]
```diff
@@ -51,47 +51,50 @@ boost::uuids::uuid CoreFbPointCloud::addData(const seerep::fb::PointCloud2& pcl)
...
 
   auto hdf5io = CoreFbGeneral::getHdf5(uuidProject, m_seerepCore, m_hdf5IoMap);
 
-  hdf5io->writePointCloudBoundingBoxLabeled(boost::lexical_cast<std::string>(uuidMsg), boundingBoxeslabeled.labels_bb());
+  hdf5io->writeLabelsFb(boost::lexical_cast<std::string>(uuidMsg), datasetUuidLabel.datasetUuid()->str(),
+                        datasetUuidLabel.labels());
 
-  std::unordered_map<std::string, std::vector<seerep_core_msgs::LabelWithInstance>> labelWithInstancePerCategory;
-  for (auto bbWithCategory : *boundingBoxeslabeled.labels_bb())
+  // TODO this is the same code as in core_fb_image::addLabel
```
Comment (jarkenau) [inline seerep_srv/seerep_core_fb/src/core_fb_pointcloud.cpp:65]: The hole method could be shared between data types. Handle it in this PR or create an Issue.

File: seerep_srv/seerep_core_pb/include/seerep_core_pb/core_pb_conversion.h [hunk 10]
```diff
@@ -38,7 +39,7 @@
 namespace seerep_core_pb
 {
 
-using LabelWithInstanceVec = std::vector<seerep_core_msgs::LabelWithInstance>;
+using LabelVec = std::vector<seerep_core_msgs::Label>;
```
Comment (jarkenau) [inline seerep_srv/seerep_core_pb/include/seerep_core_pb/core_pb_conversion.h:42]: https://github.com/agri-gaia/seerep/pull/354#discussion_r1422669544 ?

File: seerep_ros/seerep_ros_conversions_fb/include/seerep_ros_conversions_fb/conversions.h [hunk 11]
```diff
@@ -359,66 +271,66 @@ flatbuffers::Offset<seerep::fb::TransformStamped> toFlat(const geometry_msgs::Tr
...
+//                                                              flatbuffers::grpc::MessageBuilder& builder,
+//                                                              std::string label = "", std::string instanceUUID = "");
+
+// /**
+//  * @brief Converts a Flatbuffer BoundingBoxes2DLabeled message to the corresponding
+//  * ROS vision_msgs::Detection2D message
+//  * @param bb_labeled_stamped Flatbuffer BoundingBoxes2DLabeled
+//  * @return ROS vision_msgs::Detection2D
+//  */
+// vision_msgs::Detection2D toROS(const seerep::fb::BoundingBox2DLabeled& bb_labeled_stamped);
```
Comment (jarkenau) [inline seerep_ros/seerep_ros_conversions_fb/include/seerep_ros_conversions_fb/conversions.h:333]: Delete?

File: seerep_hdf5/seerep_hdf5_core/include/seerep_hdf5_core/impl/hdf5_core_general.hpp [hunk 12]
```diff
@@ -88,4 +88,14 @@ void Hdf5CoreGeneral::readHeader(const std::string& id, HighFive::AnnotateTraits
   header.frameId = Hdf5CoreGeneral::readFrameId(object, HEADER_FRAME_ID, id);
   header.sequence = Hdf5CoreGeneral::readAttributeFromHdf5<int32_t>(object, HEADER_SEQ, id);
 }
+
+template <class T>
+T Hdf5CoreGeneral::readDataset(const std::string& path)
+{
+  checkExists(path);
```
Comment (jarkenau) [inline seerep_hdf5/seerep_hdf5_core/include/seerep_hdf5_core/impl/hdf5_core_general.hpp:95]: Is marked as deprecated!

File: seerep_hdf5/seerep_hdf5_core/include/seerep_hdf5_core/hdf5_core_general.h [hunk 13]
```diff
@@ -147,23 +148,23 @@ class Hdf5CoreGeneral
...
+  //  */
+  // void readBoundingBoxLabeledAndAddToLabelsWithInstancesWithCategory(
+  //     const std::string& datatypeGroup, const std::string& uuid,
+  //     std::unordered_map<std::string, std::vector<seerep_core_msgs::LabelWithInstance>>& labelsWithInstancesWithCategory);
+  // void readBoundingBoxLabeled(const std::string& datatypeGroup, const std::string& uuid,
+  //                             std::vector<std::string>& labelCategories,
+  //                             std::vector<std::vector<std::string>>& labelsPerCategory,
+  //                             std::vector<std::vector<float>>& labelConfidencesPerCategory,
+  //                             std::vector<std::vector<std::vector<double>>>& boundingBoxesPerCategory,
+  //                             std::vector<std::vector<std::string>>& instancesPerCategory, bool loadBoxes = true);
```
Comment (jarkenau) [inline seerep_hdf5/seerep_hdf5_core/include/seerep_hdf5_core/hdf5_core_general.h:168]: Remove?

File: seerep_msgs/core/label_category.h [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+namespace seerep_core_msgs
+{
+struct LabelCategory
+{
+  std::string category;
+  std::vector<std::string> labels;
+  std::vector<int> labelsIdDatumaro;
+  std::vector<std::string> instances;
+  std::vector<int> instancesIdDatumaro;
+  std::string datumaroJson;
```
Comment (jarkenau) [inline seerep_msgs/core/label_category.h:15]: Why is this not a `std::vector<Label>` from `label.h`? This way, we would also not rely on an index based matching of related data.

Comment (jarkenau) [reply seerep_msgs/core/label_category.h:15]: Ah, I figured out why, this way the automatic conversion of `std::vector<>` to `datasets` from HighFive can be leveraged. See here: https://github.com/agri-gaia/seerep/blob/dd983f9cfbdf433ef766cb08a16d7f7d7aaf40f3/seerep_hdf5/seerep_hdf5_core/src/hdf5_core_general.cpp#L156-L180
However, I still think it's not a good approach.

Comment (Mark-Niemeyer) [reply seerep_msgs/core/label_category.h:15]: I agree, but we have to copy the data in such a structure at one point anyway. Thus, we can do it directly when we copy the data the first time and avoid copying it a second time. Maybe there are better approaches, but for the time being it works.

**Pull request merged**